packages/flutter/lib/src/cupertino/switch.dart

```diff
@@ -554,39 +576,75 @@ class _CupertinoSwitchState extends State<CupertinoSwitch>
     return trackInnerLength;
   }
 
+  void _handleOnTapDown(TapDownDetails details) {
+    if (isInteractive) {
+      _dragStartPosition = details.globalPosition;
+    }
+  }
+
   void _handleDragStart(DragStartDetails details) {
     if (isInteractive) {
       reactionController.forward();
-      _emitVibration();
+
+      if (_dragStartPosition != Offset.zero) {
+        final double delta = (details.globalPosition - _dragStartPosition).dx / _kTrackWidth;
+        _dragDelta = switch (Directionality.of(context)) {
+          TextDirection.rtl => -delta,
+          TextDirection.ltr => delta,
+        };
+      }
+
+      _dragValue = value;
     }
   }
 
   void _handleDragUpdate(DragUpdateDetails details) {
     if (isInteractive) {
-      position
-        ..curve = Curves.linear
-        ..reverseCurve = Curves.linear;
-      final double delta = details.primaryDelta! / _trackInnerLength;
-      positionController.value += switch (Directionality.of(context)) {
+      final double delta = details.primaryDelta! / _kTrackWidth;
+      _dragDelta += switch (Directionality.of(context)) {
         TextDirection.rtl => -delta,
         TextDirection.ltr => delta,
       };
+
+      final bool valueChangedWhileDragging = widget.value != _dragValue;
+
+      final double threshold = valueChangedWhileDragging
+          ? _kDragReverseThreshold
+          : _kDragCommitThreshold;
+      final double effectiveThreshold = widget.value ? -threshold : threshold;
+
+      final bool currentDragValue = _dragDelta >= effectiveThreshold;
```

Comment:
Nit: I think this gets a little confusing on which "drag value" means what at a glance. currentDragValue may be better named to something like "newDragValue".